pkg/runcfanotify/runcfanotify.go

```diff
@@ -0,0 +1,352 @@
+	if err != nil {
+		log.Infof("Cannot ignore %s/config.json: %s\n", bundleDir, err)
+	}
+
+	containerID := filepath.Base(filepath.Clean(bundleDir))
+
+	// f returns containerID, containerPID, error
+	f := func(notify *fanotify.NotifyFD, callback RuncNotifyFunc) (bool, error) {
+		data, err := notify.GetEvent()
+		if err != nil {
+			return false, fmt.Errorf("%w", err)
+		}
+
+		if data == nil {
+			return false, nil
+		}
+
+		defer data.Close()
+		if !data.MatchMask(unix.FAN_ACCESS_PERM) {
+			return false, fmt.Errorf("fanotify: unknown event")
+		}
+		defer notify.ResponseAllow(data)
+
+		pid := data.GetPID()
+
+		// Cannot use GetEvent(skipPIDs) because we need to call ResponseAllow
+		if pid == os.Getpid() {
+			return false, nil
+		}
+
+		path, err := data.GetPath()
+		if err != nil {
+			return false, err
+		}
+		if path != pidFile {
+			return false, nil
+		}
+
+		commB, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/comm", pid))
+		if err != nil {
+			return false, err
+		}
+		comm := strings.TrimSuffix(string(commB), "\n")
+
+		if comm != "containerd-shim" {
+			return false, nil
+		}
+
+		// Unfortunately, Linux 5.4 doesn't respect ignore masks
+		// See fix in Linux 5.9:
+		// https://github.com/torvalds/linux/commit/497b0c5a7c0688c1b100a9c2e267337f677c198e
+		// Workaround: remove parent mask. We don't need it anymore :)
+		err = notify.Mark(unix.FAN_MARK_REMOVE, unix.FAN_ACCESS_PERM|unix.FAN_EVENT_ON_CHILD, unix.AT_FDCWD, pidFileDir)
+		if err != nil {
+			return false, nil
+		}
+
+		pidFileContent, err := ioutil.ReadAll(data.File())
+		if err != nil {
+			return false, err
+		}
+		containerPID, err := strconv.Atoi(string(pidFileContent))
+		if err != nil {
+			return false, err
+		}
+
+		pidfd, _, errno := unix.Syscall(unix.SYS_PIDFD_OPEN, uintptr(containerPID), 0, 0)
+		if errno != 0 {
+			return false, fmt.Errorf("pidfd_open returned %v", errno)
+		}
+
+		// Read the OCI config to find the container name in the pod
+		log.Infof("runcfanotify: ReadFile %s", filepath.Join(bundleDir, "config.json"))
+		bundleConfigJson, err := ioutil.ReadFile(filepath.Join(bundleDir, "config.json"))
+		log.Infof("runcfanotify: ReadFile %s DONE", filepath.Join(bundleDir, "config.json"))
+		if err != nil {
+			return false, err
+		}
+		containerConfig := &ocispec.Spec{}
+		err = json.Unmarshal(bundleConfigJson, containerConfig)
+		if err != nil {
+			return false, err
+		}
+
+		terminationLogRegexp := regexp.MustCompile("/kubelet/pods/([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})/containers/(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])/")
+		podUID := ""
+		containerName := ""
+		mountSources := []string{}
+		for _, m := range containerConfig.Mounts {
+			mountSources = append(mountSources, m.Source)
+			matches := terminationLogRegexp.FindStringSubmatch(m.Source)
+			if len(matches) == 4 {
+				podUID = matches[1]
+				containerName = matches[2]
+			}
+		}
+
```

Comment:
I think I'm missing something. We have a similar logic in `pkg/gadgettracermanager/k8s/k8s.go:FillContainer()`. Why we don't have a regular expression there? Is it possible to combine this logic and the one in `FillContainer`?

Reply:
You're right, it should be in one place.

I would like the `podUID` to be set in the runcfanotify package so it could be conveniently used by external projects. At the same time, other hook modes (`gadget-container/hooks/oci/main.go`)  are doing the same thing. So I am not sure where is the best place.